Sensible_Ambushing/Mods/Sensible_Ambushing/ScriptExtender/Lua/Extras/_StealthAttacks.lua

```diff
@@ -0,0 +1,212 @@
+	if spellProperties and next(stealth_tracker) then
+		for _, functorGroup in pairs(spellProperties) do
+			for _, functor in pairs(functorGroup.Functors) do
+				-- ApplyStatus check needs to come first, otherwise we'll get a runtime exception for StatusId missing
+				if functor.TypeId == "ApplyStatus" and functor.StatusId == "SNEAKING" then
+					Logger:BasicTrace("%s cast %s, which applies sneaking, so removing the stealth tracker", caster, spell)
+					caster_entity.Vars.Sensible_Ambushing_Stealth_Action_Tracker = nil
+					return
+				end
+			end
+		end
+	end
+
+	if Osi.HasActiveStatus(caster, "SNEAKING") == 1 then
+		if Osi.SpellHasSpellFlag(spell, "Stealth") == 0 then
+			stealth_tracker.Counter = (stealth_tracker.Counter or 0) + 1
+
+			if Osi.IsInCombat(caster) == 0 then
+				Ext.Timer.WaitFor(3000, function()
+					-- cuz caching - https://github.com/Norbyte/bg3se/blob/main/Docs/API.md#caching-behavior
+					if caster_entity.Vars.Sensible_Ambushing_Stealth_Action_Tracker and Osi.IsInCombat(caster) == 0 then
+						Logger:BasicDebug(
+							"%s acted from stealth and was eligible to stealth attack, but they didn't enter combat, so removing tracker and reapplying SNEAKING if they're still missing it",
+							caster)
+						caster_entity.Vars.Sensible_Ambushing_Stealth_Action_Tracker = nil
+						if Osi.HasActiveStatus(caster, "SNEAKING") == 0 then
+							Osi.ApplyStatus(caster, "SNEAKING", -1, 0)
+						end
+					end
+				end)
+			else
+				RollStealthAgainstEnemies(caster)
+			end
+		end
+	end
+	caster_entity.Vars.Sensible_Ambushing_Stealth_Action_Tracker = stealth_tracker
+end)
+
+EventCoordinator:RegisterEventProcessor("CombatStarted", function(combatGuid)
+	for _, ambushingCombatMember in pairs(Osi.DB_Is_InCombat:Get(nil, combatGuid)) do
+		ambushingCombatMember = ambushingCombatMember[1]
+		local entity = Ext.Entity.Get(ambushingCombatMember)
+		local stealth_tracker = entity.Vars.Sensible_Ambushing_Stealth_Action_Tracker or {}
+
+		if stealth_tracker.Counter then
+			-- Just in case there's some gap in events and it doesn't get reset
+			stealth_tracker.Counter = 1
+			Logger:BasicDebug("%s acted from stealth and should keep their stealth in combat, so doing that", ambushingCombatMember)
+
+			Osi.ApplyStatus(ambushingCombatMember, "SNEAKING", -1, 0)
+
+			RollStealthAgainstEnemies(ambushingCombatMember)
+
+			entity.Vars.Sensible_Ambushing_Stealth_Action_Tracker = stealth_tracker
+			return
+		end
+	end
+end)
+
+EventCoordinator:RegisterEventProcessor("RollResult", function(eventName, stealthActor, enemy, resultType, _, criticality)
+	if eventName == "Sensible_Ambush_Stealth_Action_Check_" .. ModuleUUID then
+		Logger:BasicTrace("Processing Ambush Stealth Action check for %s against %s with result %s and criticality %s",
+			stealthActor,
+			enemy,
+			resultType,
+			criticality)
+
+		local entity = Ext.Entity.Get(stealthActor)
+		local stealth_tracker = entity.Vars.Sensible_Ambushing_Stealth_Action_Tracker or {}
```

Comment:
Note to self - don't default here. Should always have it, or if we don't then we need to cancel this